KAFKA-13063: Make DescribeConsumerGroupsHandler unmap for COORDINATOR_NOT_AVAILABLE error

showuon opened the pull request:
refactor DescribeConsumerGroupsHandler and tests. Also, put `COORDINATOR_NOT_AVAILABLE` as unmap retry.

the old handleResponse for `DescribeConsumerGroups` request:
```java
void handleResponse(AbstractResponse abstractResponse) {
      final DescribeGroupsResponse response = (DescribeGroupsResponse) abstractResponse;

      List<DescribedGroup> describedGroups = response.data().groups();
      if (describedGroups.isEmpty()) {
          context.future().completeExceptionally(
                  new InvalidGroupIdException("No consumer group found for GroupId: " + context.groupId()));
          return;
      }

      if (describedGroups.size() > 1 ||
              !describedGroups.get(0).groupId().equals(context.groupId())) {
          String ids = Arrays.toString(describedGroups.stream().map(DescribedGroup::groupId).toArray());
          context.future().completeExceptionally(new InvalidGroupIdException(
                  "DescribeConsumerGroup request for GroupId: " + context.groupId() + " returned " + ids));
          return;
      }

      final DescribedGroup describedGroup = describedGroups.get(0);

      // If coordinator changed since we fetched it, retry
      if (ConsumerGroupOperationContext.hasCoordinatorMoved(response)) {
          Call call = getDescribeConsumerGroupsCall(context);
          rescheduleFindCoordinatorTask(context, () -> call, this);
          return;
      }

      final Errors groupError = Errors.forCode(describedGroup.errorCode());
      if (handleGroupRequestError(groupError, context.future()))
          return;

      final String protocolType = describedGroup.protocolType();
      if (protocolType.equals(ConsumerProtocol.PROTOCOL_TYPE) || protocolType.isEmpty()) {
          final List<DescribedGroupMember> members = describedGroup.members();
          final List<MemberDescription> memberDescriptions = new ArrayList<>(members.size());
          final Set<AclOperation> authorizedOperations = validAclOperations(describedGroup.authorizedOperations());
          for (DescribedGroupMember groupMember : members) {
              Set<TopicPartition> partitions = Collections.emptySet();
              if (groupMember.memberAssignment().length > 0) {
                  final Assignment assignment = ConsumerProtocol.
                      deserializeAssignment(ByteBuffer.wrap(groupMember.memberAssignment()));
                  partitions = new HashSet<>(assignment.partitions());
              }
              final MemberDescription memberDescription = new MemberDescription(
                      groupMember.memberId(),
                      Optional.ofNullable(groupMember.groupInstanceId()),
                      groupMember.clientId(),
                      groupMember.clientHost(),
                      new MemberAssignment(partitions));
              memberDescriptions.add(memberDescription);
          }
          final ConsumerGroupDescription consumerGroupDescription =
              new ConsumerGroupDescription(context.groupId(), protocolType.isEmpty(),
                  memberDescriptions,
                  describedGroup.protocolData(),
                  ConsumerGroupState.parse(describedGroup.groupState()),
                  context.node().get(),
                  authorizedOperations);
          context.future().complete(consumerGroupDescription);
      } else {
          context.future().completeExceptionally(new IllegalArgumentException(
              String.format("GroupId %s is not a consumer group (%s).",
                  context.groupId(), protocolType)));
      }
  }
```


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon committed:
KAFKA-13063: refactor DescribeConsumerGroupsHandler and tests

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeConsumerGroupsHandler.java:
```diff
@@ -103,6 +104,12 @@ public String apiName() {
+        if (describedGroups.isEmpty()) {
+            throw new InvalidGroupIdException("No consumer group found");
+        }
+    }
```
What's the purpose of this check? I am not sure where this `InvalidGroupIdException` thrown here will get to.

showuon replied:
Before KIP-699, there's the check:
```java
if (describedGroups.isEmpty()) {
          context.future().completeExceptionally(
                  new InvalidGroupIdException("No consumer group found for GroupId: " + context.groupId()));
          return;
      }
```
But, you're right, the exception thrown will not return back to requester. Also, because there's no group found, we cannot have a failed key return because we need a group in key. 

Removed it since it should never happen.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeConsumerGroupsHandler.java:
```diff
@@ -151,38 +162,59 @@ public String apiName() {
         }
-        return new ApiResult<>(completed, failed, unmapped);
+
+        if (groupsToUnmap.isEmpty() && groupsToRetry.isEmpty()) {
```
It seems incorrect here as well.

showuon replied:
You're right! We accept multiple groups. Updated. Thanks.

showuon committed:
KAFKA-13063: refactor codes

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-13063

showuon committed:
KAFKA-13063: refactor

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeConsumerGroupsHandler.java:
```diff
@@ -109,16 +109,19 @@ public String apiName() {
             Errors error = Errors.forCode(describedGroup.errorCode());
             if (error != Errors.NONE) {
-                handleError(groupIdKey, error, failed, unmapped);
+                handleError(groupIdKey, error, failed, groupsToUnmap, groupsToRetry);
```
`groupsToRetry` to retry is not really necessary in this case. We don't even use it later. Could we remove it?

showuon replied:
Good catch! Updated.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeConsumerGroupsHandler.java:
```diff
@@ -151,38 +154,45 @@ public String apiName() {
         }
-        return new ApiResult<>(completed, failed, unmapped);
+
+        return new ApiResult<>(completed, failed,  new ArrayList<>(groupsToUnmap));
```
nit: There is an extra space before `new`.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeConsumerGroupsHandler.java:
```diff
@@ -151,38 +154,45 @@ public String apiName() {
+                final String unexpectedErrorMsg =
+                    String.format("`DescribeGroups` request for group id %s failed due to error %s", groupId.idValue, error);
+                log.error(unexpectedErrorMsg);
+                failed.put(groupId, error.exception(unexpectedErrorMsg));
```
We don't provide the error message in any other case. Should we remove this one for the time being? I think that it is a good idea but only if we do it across the board.

showuon replied:
Agree. I just follow the previous behavior. Remove the error message. Thanks.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/DescribeConsumerGroupsHandler.java:
```diff
@@ -199,4 +209,4 @@ private void handleError(
     }
 
-}
+}
```
nit: Could we revert this back?

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -2688,8 +2688,7 @@ public void testDescribeConsumerGroups() throws Exception {
             env.kafkaClient().setNodeApiVersions(NodeApiVersions.create());
 
-            //Retriable FindCoordinatorResponse errors should be retried
-            env.kafkaClient().prepareResponse(prepareFindCoordinatorResponse(Errors.COORDINATOR_NOT_AVAILABLE,  Node.noNode()));
```
Do we really need to remove this one? It seems to me that the changes in the PR does not change how the find coordinator response is handled, no?

showuon replied:
Oh, you are right! reverted my change. Thank you!

dajac reviewed (commented):
@showuon Thank for the update. I left a few more comments.

showuon committed:
KAFKA-13063: revert find coordinator response change

showuon commented:
@dajac , I addressed your comments in this PR and all other 4 PRs. And also update the PR title and description accordingly. Please help take a look when available. Thank you.

dajac reviewed (approved):
LGTM

dajac commented:
Failures are not related:
```
Build / JDK 8 and Scala 2.12 / testUnauthorizedDeleteTopicsWithDescribe() – kafka.api.AuthorizerIntegrationTest
1m 7s
Build / JDK 8 and Scala 2.12 / testCloseDuringRebalance() – kafka.api.ConsumerBounceTest
18s
Build / JDK 8 and Scala 2.12 / testListReassignmentsDoesNotShowNonReassigningPartitions() – kafka.api.PlaintextAdminIntegrationTest
5s
Build / JDK 8 and Scala 2.12 / testSendOffsetsToTransactionTimeout() – kafka.api.TransactionsTest
```

dajac commented:
Merged to trunk and to 3.0. cc @kkonstantine
